File: algorithms/leetcode/238-product-of-array-except-self/product-of-array-except-self.cpp
#include <vector>

class Solution {
public:
  std::vector<int> productExceptSelf(std::vector<int>& nums) {
    std::vector<int> result(nums.size(), 1);

    for (std::vector<int>::size_type i = 0; i < nums.size() - 1; i++) {
      result[i + 1] = result[i] * nums[i];
    }

    int product = 1;
    for (std::vector<int>::size_type i = nums.size(); i > 0; i--) {
      result[i - 1] *= product;
      product *= nums[i - 1];
    }

    return result;
  }
};

File: leetcode/238-product-of-array-except-self/product-of-array-except-self.js

const productExceptSelf = nums => {
    const result = [1];

    for (let i = 1; i < nums.length; i++) {
        result.push(result[i - 1] * nums[i - 1]);
    }

    for (let i = nums.length - 1, add = 1; i >= 0; i--) {
        result[i] *= add;
        add *= nums[i];
    }

    return result;
}
